llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,322 @@
+        if (PV->PageToDataIdx[PageIdx] == InvalidPage)
+          continue;
+
+        Count += PAGE_SIZE;
+      }
+      Count += ElementMax % PAGE_SIZE;
+      Count += PAGE_SIZE - ElementMin % PAGE_SIZE;
+
+      return Count;
+    }
+
+    // When dereferencing the iterator, we materialise the page if needed.
+    T const &operator*() const {
+      assert(ElementIdx < PV->Size);
+      assert(PV->PageToDataIdx[ElementIdx / PAGE_SIZE] != InvalidPage);
+      T *PagePtr =
+          reinterpret_cast<T *>(PV->PageToDataIdx[ElementIdx / PAGE_SIZE]);
+      return *((ElementIdx % PAGE_SIZE) + PagePtr);
+    }
+
+    // Equality operator.
+    bool operator==(MaterialisedIterator const &Other) const {
+      // Iterators of two different vectors are never equal.
+      if (PV != Other.PV)
+        return false;
+      // Any iterator for an empty vector is equal to any other iterator.
+      if (PV->empty())
+        return true;
+      // Get the pages of the two iterators. If between the two pages there
+      // are no valid pages, we can condider the iterators equal.
+      std::size_t PageMin = std::min(ElementIdx, Other.ElementIdx) / PAGE_SIZE;
+      std::size_t PageMax = std::max(ElementIdx, Other.ElementIdx) / PAGE_SIZE;
+      // If the two pages are past the end, the iterators are equal.
+      if (PageMin >= PV->PageToDataIdx.size())
+        return true;
+      // If only the last page is past the end, the iterators are equal if
+      // all the pages up to the end are invalid.
+      if (PageMax >= PV->PageToDataIdx.size()) {
+        for (std::size_t PageIdx = PageMin; PageIdx < PV->PageToDataIdx.size();
+             ++PageIdx)
+          if (PV->PageToDataIdx[PageIdx] != InvalidPage)
+            return false;
+        return true;
+      }
+
+      uintptr_t Page1 = PV->PageToDataIdx[PageMin];
+      uintptr_t Page2 = PV->PageToDataIdx[PageMax];
+      if (Page1 == InvalidPage && Page2 == InvalidPage)
+        return true;
+
+      // If the two pages are the same, the iterators are equal if they point
+      // to the same element.
+      if (PageMin == PageMax)
+        return ElementIdx == Other.ElementIdx;
+
+      // If the two pages are different, the iterators are equal if all the
+      // pages between them are invalid.
+      for (std::size_t PageIdx = PageMin; PageIdx < PageMax; ++PageIdx)
+        if (PV->PageToDataIdx[PageIdx] != InvalidPage)
+          return false;
+
+      return true;
+    }
+
+    bool operator!=(MaterialisedIterator const &Other) const {
+      return !(*this == Other);
+    }
+
+    [[nodiscard]] size_t getIndex() const { return ElementIdx; }
+  };
+
+  // Iterators over the materialised elements of the vector.
+  // This includes all the elements belonging to allocated pages,
+  // even if they have not been accessed yet. It's enough to access
+  // one element of a page to materialise all the elements of the page.
+  MaterialisedIterator materialisedBegin() const {
```

Comment:
`materialized_begin` to conform to the coding style: https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly